drivers/sftp/onitu_sftp/sftp.py

```diff
@@ -0,0 +1,325 @@
+    try:
+        create_file_subdirs(sftp, metadata.path)
+        plug.logger.debug(u"Adding {} to ignored files".format(metadata.path))
+        events_to_ignore.add(metadata.path)
+    except ServiceError:
+        raise
+    finally:
+        sftp.close()
+
+
+@plug.handler()
+def upload_file(metadata, content):
+    plug.logger.debug(u"Uploading file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        f = sftp.open(metadata.path, 'w+')
+        f.write(content)
+        f.close()
+    except (IOError, OSError) as e:
+        raise ServiceError("Error writing file '{}': {}"
+                           .format(metadata.path, e))
+    finally:
+        sftp.close()
+
+
+@plug.handler()
+def end_upload(metadata):
+    plug.logger.debug(u"Ending upload of file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        stat_res = sftp.stat(metadata.path)
+        update_metadata(metadata, stat_res.st_size, stat_res.st_mtime,
+                        update_file=False)
+    except (IOError, OSError) as e:
+        raise ServiceError(u"Error during upload ending on file '{}': {}"
+                           .format(metadata.path, e))
+    finally:
+        sftp.close()
+        if metadata.path in events_to_ignore:
+            plug.logger.debug(u"Removing {} to ignored files"
+                              .format(metadata.path))
+            events_to_ignore.remove(metadata.path)
+
+
+@plug.handler()
+def abort_upload(metadata):
+    plug.logger.debug(u"Aborting upload of file {}".format(metadata.path))
+    delete_file(metadata)
+
+
+@plug.handler()
+def delete_file(metadata):
+    plug.logger.debug(u"Deleting file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        sftp.remove(metadata.path)
+    except (IOError, OSError) as e:
+        raise ServiceError(u"Error deleting file '{}': {}"
+                           .format(metadata.path, e))
+    finally:
+        if metadata.path in events_to_ignore:
+            plug.logger.debug(u"Removing {} to ignored files"
+                              .format(metadata.path))
+            events_to_ignore.remove(metadata.path)
+        sftp.close()
+
+
+@plug.handler()
+def move_file(old_metadata, new_metadata):
+    plug.logger.debug(u"Moving file {} to {}"
+                      .format(old_metadata.path, new_metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        sftp.stat(new_metadata.path)
+    except IOError:  # file doesn't exist: good
+        try:
+            create_file_subdirs(sftp, new_metadata.path)
+            sftp.rename(old_metadata.path, new_metadata.path)
+        except IOError as ioe:
+            raise ServiceError(u"Error while moving file {} to {}: {}"
+                               .format(old_metadata.path, new_metadata.path,
+                                       ioe))
+    else:
+        raise ServiceError(u"File {} already exists !"
```

Comment:
Agreed, I got mixed up with Paramiko's SFTP server interface's [rename](http://docs.paramiko.org/en/latest/api/sftp.html#paramiko.sftp_si.SFTPServerInterface.rename), for which the Paramiko documentation advises to treat rename of an existing file as an error.